src/FreeAge/common/type_info.cpp

```diff
@@ -0,0 +1,621 @@
+///
+/// CRITICAL: Also here are all the information about values that are needed for a single unit.
+///
+///           Example: villagerCarryingCapacity: an obvius design would be to add this
+///           to the villager unit type that will contain this. Because only the Villager need
+///           this, it would introduce some variation in the UnitTypeStats, like inheritance or
+///           a list of generic attribute objects. The other solution is to put in the CivilizationStats.
+///
+///           Problems solved by keeping the UnitTypeStats as simple struct without pointers and lists.
+///           - Can be placed inside a vector<UnitTypeStats> (without pointer) = memory management done
+///           - The number of UnitTypeStats is known at start + a properly sized vector =
+///             simple pointers to UnitTypeStats can be used safely
+///           - Huge reduction of the code base.
+///           - Cleaner implemention of the gamelogic.
+///           - Simpler information retrieval for the gui.
+///           - Performance (size, locality, less indirection)
+///
+///
+struct CivilizationStats {
+
+  // Add to the starting resource of the player
+  ResourceAmount startingResourceD;
+
+  UnitType startingScoutUnit;
+  int startingVillagerCount;
+  bool startingLlama;
+
+  int populationCap;
+  int bonusPopulationSpace;
+
+  float garrisonHealRate;
+
+  // TODO: monk conversion things
+
+  float tributeFee;
+  float tradingFree;
+  float tradeGoldGeneration;
+  float relicGoldGeneration;
+
+  // == single unit values
+
+  int farmFood;
+
+  // NOTE: could aslo be stored as the workRate of an unspecialized villager
+  float villagerConstuctionSpeed;
+
+  float villagerCarryingCapacity[/* TODO: number of villager types*/];
+  float fishingShipCarryingCapacity;
+
+  bool singleTowerCenter;
+  // use singleTowerCenter this instead of a max limit for every building (only the town center has
+  // max limit, even the wonder can be build multiple types)
+
+};
+
+/// Represents a modification to something
+struct Modification {
+
+  static constexpr i32 Unset = std::numeric_limits<i32>::min();
+
+  enum class Type {
+    // unit, building and technology
+    Cost,            // use extra to define ResourceType or leave Unset for to all of them
+
+    // unit and building
+    MaxHp,
+    Damage,          // use extra to define DamageType
+    Armor,           // use extra to define DamageType
+    MinRange,
+    MaxRange,
+    FireRate,
+    Accuracy,        // use extra to define type of accuracy ?
+    LineOfSight,
+    GarrisonCapacity,
+    Upgrade,         // option 1: use value to define the old type, use extra to define the new unit (applied to civ)
+                     // option 2: use value to define the new unit (applied to unit)
+
+    // unit only
+    Speed,
+    ProductionTime,
+
+    // building only
+    ConstructionSpeed,
+    WorkRate,
+    PopulationSpace,
+
+    // technology only
+    ResearchTime,
+
```

Comment:
Add a `MakeFree` instead of setting the cost to zero, should be a different type because it does not only describe the cost but also that the technology is automatically researched.

And the `TechnologyStats` could have a corresponding:
```c++
  // ...
  enum Availability {
    Normal, FreeFromTheStart, FreeFromWhenTheAgeRequirmentIsMet
  }
  Availability availability
  // ...
```

Edit: `MakeFree` could be renamed to `Availability` and use the value to pass one of the `TechnologyStats::Availability`